libavfilter/dnn/dnn_backend_tf.c

```diff
@@ -197,6 +198,11 @@ static DNNReturnType set_input_output_tf(void *model, DNNData *input, const char
     }
 
     sess_opts = TF_NewSessionOptions();
+    // protobuf data for auto memory gpu_options.allow_growth=True and gpu_options.visible_device_list="0" 
```

Comment:
Why do we need the `gpu_options.allow_growth=True` flag?

And, based on the name, it sounds like `gpu_options.visible_device_list="0"` would only provide access to the GPU with device ID 0 to the program. What if we want to use a GPU on the machine that does not have device ID 0?

Reply:
By default, TensorFlow maps nearly all of the GPU memory of all GPUs visible to the process. This is done to more efficiently use the relatively GPU memory resources on the devices by reducing memory fragmentation. 
In our case, it is desirable for the process to only allocate a subset of the available memory, or to only grow the memory usage as is needed by the process. So I used gpu_options.allow_growth=True flag.
In the multi GPU, I think we need to test some more.

Reply:
Thanks for the explanation on why we're using `gpu_options.allow_growth=True`!

> In the multi GPU, I think we need to test some more.

Would it be possible to use the device specified by LPMS? For example, LPMS [configures the scale filter for an output based on the device specified](https://github.com/livepeer/lpms/blob/6e5c36facc442d4474de425d79038aec6d293dcc/ffmpeg/ffmpeg.go#L224).

Reply:
Bump on this. Is there a way to avoid setting `gpu_options.visible_device_list="0"`? If I understand correctly, this option would restrict the filter to running on the GPU #0. It'd be preferable if the filter can be run on the GPU that transcoding is executed on which could have any device ID.

Reply:
If I understand the comments [here](https://github.com/tensorflow/tensorflow/blob/801e66418b63368bcff99a98ce650b3b34daf5e9/tensorflow/core/protobuf/config.proto#L78) correctly, we can tell TF to only use GPU N by setting `gpu_options.visible_device_list=N` and then within the TF process GPU N would be visible as "/device:GPU:0". If this is the case, then if we want to tell the TF process what GPU to use we need a way to dynamically set `gpu_options.visible_device_list` instead of hardcoding its value like we do now.

From a review of the ffmpeg TF code, I *think* we can dynamically set `visible_device_list` when the detection filter is initialized.

Currently, we load the TF model when the detection filter is initialized on this line:

```c
ctx->model = (ctx->dnn_module->load_model)(ctx->model_filename);
```

Since we're using the TF backend, `load_module` is [set to](https://github.com/FFmpeg/FFmpeg/blob/918fc9a0ed4e9202341ffb3e7e4b72b4387dfe0a/libavfilter/dnn/dnn_interface.c#L49) [ff_dnn_load_model_tf] which accepts an `options` argument which is used to configure the [TF model's context](https://github.com/FFmpeg/FFmpeg/blob/918fc9a0ed4e9202341ffb3e7e4b72b4387dfe0a/libavfilter/dnn/dnn_backend_tf.c#L710). The `load_tf_model` function then uses the [sess_config](https://github.com/FFmpeg/FFmpeg/blob/918fc9a0ed4e9202341ffb3e7e4b72b4387dfe0a/libavfilter/dnn/dnn_backend_tf.c#L240) in the model's context to configure the TF session.

I couldn't find the docs on this, but if `sess_config` represents the same data as the [ConfigProto](https://github.com/tensorflow/tensorflow/blob/801e66418b63368bcff99a98ce650b3b34daf5e9/tensorflow/core/protobuf/config.proto#L380) message, then I think we can set `gpu_options.visible_device_list` in `sess_config` to tell the TF session which GPU to use.